### DISCUSSION TITLE
Are FINN DMA the same as vivado AXI_DMA ?

### DISCUSSION THREAD

**Discussion by kalahel** [Xilinx/finn-hlslib#145] [General]
Hi

I'm currently trying to write custom C driver code for a FINN project but I struggle a bit to interact and understand DMA generated by FINN.

Do they they wrap classic Vivado DMA or are they entirely custom ? Are they using the same control registers ?

Best regards,
Mat

**Comment by auphelia** [comment]
Hi @kalahel,
Another user opened a PR with this a while ago: https://github.com/Xilinx/finn/pull/989
Maybe that could be helpful, they're using xrt for the C++ driver.

**Comment by fpjentzsch** [comment]
Hi,

the DMA is a custom HLS implementation [here](https://github.com/Xilinx/finn-hlslib/blob/master/dma.h), which is instantiated [here](https://github.com/Xilinx/finn/blob/8da09eb5b3567a4c466b06bd340393823985784b/src/finn/custom_op/fpgadataflow/hls/iodma_hls.py#L224) with interface directives applied [here](https://github.com/Xilinx/finn/blob/8da09eb5b3567a4c466b06bd340393823985784b/src/finn/custom_op/fpgadataflow/hls/iodma_hls.py#L367). The FINN-generated driver interacts with the resulting control registers [here](https://github.com/Xilinx/finn/blob/8da09eb5b3567a4c466b06bd340393823985784b/src/finn/qnn-data/templates/driver/driver_base.py#L337).

You could also use a normal Vivado DMA with your custom driver if you wish.

**Comment by kalahel** [comment]
Hi,

Thank you both for your very precise anwsers !

**Comment by kalahel** [comment]
Hi,

I previously ran this [end to end finn notebook](https://github.com/Xilinx/finn/blob/main/notebooks/end2end_example/bnn-pynq/tfc_end2end_example.ipynb), everything went fine, now I got a vivado project and the pynq driver generated.

I modified the vivado project and tried using classic (micro) DMA IP to interact with the neural network IP.
To accommodate for the input/output size of the model I used `AXI 4 Stream data width converter IP` as such :
![image](https://github.com/user-attachments/assets/b5141cf2-24cd-4dc9-823b-e969d30d173a)

(So 32 bits from the DMA -> 392 from the data width converter -> 8 bits from the neural network IP -> 32 bits from the converter to the DMA)

The model does not have a `tlast` port so I connected its `tvalid` port to the `tlast` input of the converter, is this a mistake ?

When I send the 784 data (28*28 sample data from MNIST ), I have nothing in output of the model.

The code I use to control the DMA is inspired from [lauri võsandi guide](https://lauri.võsandi.com/hdl/zynq/xilinx-dma.html).

Do you have a idea of what can cause the issue ?

**Reply by fpjentzsch** [reply]
Of course a 32 bit wide input DMA might bottleneck your model if it is parallelized to a 392 bit input width, but it should still work.

If your model outputs only one 8 bit value per frame/classification, connecting `tlast` to `tvalid` should be fine. Otherwise, you might need to look at the [InsertTLastMarker](https://github.com/Xilinx/finn/blob/ac2fa5a6b86146ff33f563dbdf5bcdcf47e3e4c2/src/finn/transformation/fpgadataflow/insert_tlastmarker.py#L37) transformation if your output DMA requires `tlast` to function.

You probably don't need the AXI4-Stream Data Width Converter on the output. Simply connect 8 of the 32 bits and leave the others dangling. Although it should still work in your case, even if you only run classification for a single frame, because the documentation states:
> Partial data may be flushed when either the TLAST bit is received or TID / TDEST changes before enough data is accumulated to send out a complete beat.

**Reply by kalahel** [reply]
Thanks for your reply !

I tried adding [InsertTLastMarker](https://github.com/Xilinx/finn/blob/ac2fa5a6b86146ff33f563dbdf5bcdcf47e3e4c2/src/finn/transformation/fpgadataflow/insert_tlastmarker.py#L37) but I struggle to find where to exactly.

I tried adding it [here](https://github.com/Xilinx/finn/blob/9d299689f2ec0895f208b8bfe3bcdcf6f450181a/notebooks/end2end_example/bnn-pynq/tfc_end2end_example.ipynb#L461) but I get this resulting graph :
![image](https://github.com/user-attachments/assets/9747ea88-e40e-44e2-b5b0-5b43d7279c46)

But that does not really seems to make sense.

**Reply by fpjentzsch** [reply]
Looks alright to me. The TLastMarker_hls node will not modify your data, but simply count the number of transactions and assert `tlast` on the output stream once a frame is complete. In your case this will be for every transaction since your output shape is (1,1).

Although... I have not used this feature in a while and I'm not sure which settings for `InsertTLastMarker(both=False, external=True, dynamic=True)` would be correct for use with the default Vivado AXI DMA. I assume
`both=False` _(only insert tlast for the output)_
`external=True` _(?)_
`dynamic=False` _(since you don't want to set the frame size via AXI-lite everytime you initialize the accelerator)_
but maybe @auphelia can chime in.

**Reply by kalahel** [reply]
by inserting as my previous message, I get an error : 
```
AssertionError: Insert tlast marker should be done
                        after inserting the FIFOs
```
So I tried inserting it later after the FIFO creation and before this [line](https://github.com/Xilinx/finn/blob/9d299689f2ec0895f208b8bfe3bcdcf6f450181a/notebooks/end2end_example/bnn-pynq/tfc_end2end_example.ipynb#L818), but I got this error :
```shell
  File "/mnt/c/Dev/finn/src/finn/transformation/fpgadataflow/insert_tlastmarker.py", line 60, in apply
    num_iters = int(custom_op.get_number_output_values())
AttributeError: 'StreamingDataflowPartition' object has no attribute 'get_number_output_values'
```

Finally I even tried inserting it in make_zynq_proj.py after this [line](https://github.com/Xilinx/finn/blob/9d299689f2ec0895f208b8bfe3bcdcf6f450181a/src/finn/transformation/fpgadataflow/make_zynq_proj.py#L351) and got the same error : 
```shell
  File "/mnt/c/Dev/finn/src/finn/transformation/fpgadataflow/insert_tlastmarker.py", line 60, in apply
    num_iters = int(custom_op.get_number_output_values())
AttributeError: 'StreamingDataflowPartition' object has no attribute 'get_number_output_values'
```

I am a bit lost on where exactly should I add this transformation.

**Reply by fpjentzsch** [reply]
What if you insert it after [this line](https://github.com/Xilinx/finn/blob/9d299689f2ec0895f208b8bfe3bcdcf6f450181a/src/finn/transformation/fpgadataflow/make_zynq_proj.py#L340) under the condition that it only applies to the middle StreamingDataflowPartition (the accelerator) and not the IDMA and ODMA partition?

If you want to use other DMAs, you shouldn't need the ZynqBuild transformation but stop at CreateStitchedIP. In general, I would recommend to use FINN's [builder flow](https://github.com/Xilinx/finn/blob/ac2fa5a6b86146ff33f563dbdf5bcdcf47e3e4c2/src/finn/builder/build_dataflow.py#L105) because it is more structured than the old end2end examples that call all transformations individually.

**Reply by kalahel** [reply]
Inserting it after [this line](https://github.com/Xilinx/finn/blob/9d299689f2ec0895f208b8bfe3bcdcf6f450181a/src/finn/transformation/fpgadataflow/make_zynq_proj.py#L340) with this code :
```python
kernel_model = kernel_model.transform(InsertFIFO())
# User code start here
if ("dma" not in prefix) and ("StreamingDataflowPartition_1" in prefix):
  InsertTLastMarker(dynamic=False)
  print(f"Inserting tlast marker into {prefix}")
  # User code end here
kernel_model = kernel_model.transform(SpecializeLayers(self.fpga_part))
```
Did not add a tlast port :
![image](https://github.com/user-attachments/assets/e1af3817-9b7f-48b0-9d80-d09fb155c187)
And produced those warnings : 
```shell
/mnt/c/Dev/finn/src/finn/transformation/fpgadataflow/create_stitched_ip.py:290: UserWarning: First node is not StreamingFIFO or IODMA.
                You may experience incorrect stitched-IP rtlsim or hardware
                behavior. It is strongly recommended to insert FIFOs prior to
                calling CreateStitchedIP.
                
```

Starting from the [builder flow tutorial](https://github.com/Xilinx/finn/blob/9d299689f2ec0895f208b8bfe3bcdcf6f450181a/tutorials/fpga_flow/build.py#L40) I don't know what should I modify/add to add a tlast port to the stitched IP since I don't find any reference to it in the [DataflowBuildConfig](https://github.com/Xilinx/finn/blob/main/src/finn/builder/build_dataflow_config.py#L149) class.
